File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/src/main/kotlin/com/runanywhere/sdk/foundation/bridge/CppBridge.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * Central coordinator for all C++ bridge operations.
 * Follows iOS CppBridge.swift architecture with two-phase initialization.
 */

package com.runanywhere.sdk.foundation.bridge

import com.runanywhere.sdk.features.TTS.System.SystemTTSModule
import com.runanywhere.sdk.foundation.bridge.CppBridge.initialize
import com.runanywhere.sdk.foundation.bridge.CppBridge.initializeServices
import com.runanywhere.sdk.foundation.bridge.CppBridge.shutdown
import com.runanywhere.sdk.foundation.bridge.CppBridge.shutdownSuspending
import com.runanywhere.sdk.foundation.bridge.extensions.CppBridgeDevConfig
import com.runanywhere.sdk.foundation.bridge.extensions.CppBridgeDevice
import com.runanywhere.sdk.foundation.bridge.extensions.CppBridgeEnvironment
import com.runanywhere.sdk.foundation.bridge.extensions.CppBridgeLLM
import com.runanywhere.sdk.foundation.bridge.extensions.CppBridgePlatformAdapter
import com.runanywhere.sdk.foundation.bridge.extensions.CppBridgeSDKEvents
import com.runanywhere.sdk.foundation.bridge.extensions.CppBridgeSTT
import com.runanywhere.sdk.foundation.bridge.extensions.CppBridgeState
import com.runanywhere.sdk.foundation.bridge.extensions.CppBridgeTTS
import com.runanywhere.sdk.foundation.bridge.extensions.CppBridgeTelemetry
import com.runanywhere.sdk.foundation.bridge.extensions.CppBridgeVAD
import com.runanywhere.sdk.foundation.bridge.extensions.CppBridgeVoiceAgent
import com.runanywhere.sdk.foundation.constants.SDKConstants
import com.runanywhere.sdk.httptransport.OkHttpHttpTransport
import com.runanywhere.sdk.infrastructure.logging.SDKLogger
import com.runanywhere.sdk.native.bridge.RunAnywhereBridge
import com.runanywhere.sdk.public.configuration.SDKEnvironment
import com.runanywhere.sdk.public.configuration.cEnvironment
import kotlinx.coroutines.runBlocking

/** Collects sanitized teardown failures while allowing later cleanup to run. */
internal class ShutdownFailureCollector {
    private var firstFailure: IllegalStateException? = null

    val hasFailure: Boolean
        get() = firstFailure != null

    fun record(message: String) {
        val failure = IllegalStateException(message)
        val first = firstFailure
        if (first == null) {
            firstFailure = failure
        } else {
            first.addSuppressed(failure)
        }
    }

    fun capture(
        message: String,
        action: () -> Unit,
    ): Boolean =
        try {
            action()
            true
        } catch (_: Throwable) {
            record(message)
            false
        }

    fun captureNativeShutdown(action: () -> Int): Boolean {
        val succeeded =
            try {
                action() == RunAnywhereBridge.RAC_SUCCESS
            } catch (_: Throwable) {
                false
            }
        if (!succeeded) {
            record("Native SDK shutdown failed")
        }
        return succeeded
    }

    fun throwIfAny() {
        firstFailure?.let { throw it }
    }
}

/**
 * CppBridge is the central coordinator for all C++ interop via JNI.
 *
 * Initialization follows a two-phase pattern:
 * - Phase 1 (synchronous): Core initialization including platform adapter registration
 * - Phase 2 (asynchronous): Platform service wiring, then commons-owned init orchestration
 *
 * CRITICAL: Platform adapter must be registered FIRST before any C++ calls.
 *
 * NOTE: This SDK is backend-agnostic. Backend registration (LlamaCPP, ONNX, etc.)
 * is handled by the individual backend modules, not by the core SDK.
 */
object CppBridge {
    private const val TAG = "CppBridge"
    private val logger = SDKLogger(TAG)

    @Volatile
    private var _environment: SDKEnvironment = SDKEnvironment.SDK_ENVIRONMENT_DEVELOPMENT

    private val lock = Any()

    @Volatile
    private var shutdownInProgress: Boolean = false

    @Volatile
    private var shutdownRetryRequired: Boolean = false

    /**
     * Current SDK environment.
     */
    val environment: SDKEnvironment
        get() = _environment

    /**
     * Whether Phase 1 initialization is complete.
     *
     * Delegates to [CppBridgeState] which is the canonical owner of the
     * runtime gate flags.
     */
    val isInitialized: Boolean
        get() = CppBridgeState.isInitialized

    /**
     * Whether Phase 2 services initialization is complete.
     *
     * Delegates to [CppBridgeState] which is the canonical owner of the
     * runtime gate flags.
     */
    val servicesInitialized: Boolean
        get() = CppBridgeState.servicesInitialized

    /**
     * Whether the native commons library is loaded.
     * This only indicates the core library - backend availability is separate.
     *
     * Delegates to [CppBridgeState] which is the canonical owner of the
     * runtime gate flags.
     */
    val isNativeLibraryLoaded: Boolean
        get() = CppBridgeState.nativeLibraryLoaded

    /**
     * Phase 1: Core Initialization (Synchronous, ~1-5ms, NO network calls)
     *
     * This is a fast, synchronous initialization that can be safely called from any thread,
     * including the main/UI thread. It does NOT make any network calls.
     *
     * Initializes the core SDK components in this order:
     * 1. Native Library Loading - Load core JNI library (if available)
     * 2. Platform Adapter - MUST be before C++ calls
     * 3. Logging configuration
     * 4. Events registration
     * 5. Telemetry configuration (stores credentials, no network)
     *
     * **Important:** Authentication, device registration, assignment fetch,
     * telemetry flush, and downloaded-model discovery happen in commons Phase
     * 2 (`rac_sdk_init_phase2_proto`), after [initializeServices] has wired
     * the platform callbacks/transport on a background thread.
     *
     * NOTE: Backend registration (LlamaCPP, ONNX) is NOT done here.
     * Backends are registered by the app calling LlamaCPP.register() and ONNX.register()
     * from the respective backend modules.
     *
     * Mirrors Swift SDK's initialize() which is also synchronous with no network calls.
     *
     * @param environment The SDK environment to use
     */
    fun initialize(environment: SDKEnvironment = SDKEnvironment.SDK_ENVIRONMENT_DEVELOPMENT) {
        synchronized(lock) {
            check(!shutdownInProgress) { "CppBridge shutdown is in progress" }
            check(!shutdownRetryRequired) { "CppBridge reset is required before initialization" }
            if (CppBridgeState.isInitialized) {
                return
            }

            val initStartTime = System.currentTimeMillis()

            _environment = environment

            // Try to load native library (optional - SDK works without it for non-inference features)
            tryLoadNativeLibrary()

            // CRITICAL: Register platform adapter FIRST before any C++ calls
            CppBridgePlatformAdapter.register()

            // Install the OkHttp HTTP transport BEFORE
            // any network I/O happens (device registration, model assignment
            // fetch, telemetry, auth all go through rac_http_request_*). The
            // adapter gives us the Android system trust store + proxy +
            // NetworkSecurityConfig for free and fixes the rc=77 SSL failure
            // on ~5% of devices. Safe to no-op if the native lib is missing.
            registerOkHttpTransport()

            if (CppBridgeState.nativeLibraryLoaded) {
                RunAnywhereBridge.racConfigureLogging(environment.cEnvironment)
            }

            // CRITICAL: Set environment early so CppBridgeDevice.isDeviceRegisteredCallback()
            // can determine correct behavior for production/staging modes
            CppBridgeTelemetry.setEnvironment(environment)

            // Resolve the durable identity and register device callbacks before
            // telemetry can consume it. Any KeyStore/persistence error aborts
            // initialization with the exact commons rac_result_t.
            CppBridgeDevice.register()

            initializeTelemetryManager(environment)

            // Register analytics events callback AFTER telemetry manager is initialized
            // This routes C++ events (LLM/STT/TTS) to telemetry for batching and HTTP transport
            val telemetryHandle = CppBridgeTelemetry.getTelemetryHandle()
            if (telemetryHandle != 0L) {
                CppBridgeSDKEvents.register(telemetryHandle)
            } else {
                logger.warn("Telemetry handle not available, analytics events will not be tracked")
            }

            CppBridgeState.isInitialized = true

            val initDurationMs = System.currentTimeMillis() - initStartTime
            logger.debug("Phase 1 complete in ${initDurationMs}ms ($environment)")
        }
    }

    /**
     * Initialize the C++ telemetry manager with device info.
     * Mirrors Swift SDK's CppBridge.Telemetry.initialize(environment:)
     *
     * Note: If device ID is unavailable (secure storage failure), telemetry is skipped
     * to avoid creating orphaned/duplicate device records. The app continues to function.
     */
    private fun initializeTelemetryManager(environment: SDKEnvironment) {
        try {
            if (!CppBridgeEnvironment.shouldSendTelemetry(environment)) {
                logger.debug("Telemetry disabled for $environment")
                return
            }
            // getDeviceIdCallback() may lazily initialize the persistent UUID
            val deviceId = CppBridgeDevice.getDeviceIdCallback()

            if (deviceId.isEmpty()) {
                // Skip telemetry rather than create orphaned records with a temporary ID
                logger.error(
                    "Device ID unavailable - telemetry will be disabled for this session. " + "This usually indicates secure storage is not properly initialized. " + "Ensure AndroidPlatformContext.initialize() is called before SDK initialization.",
                )
                return
            }

            val provider = CppBridgeDevice.deviceInfoProvider
            val deviceModel = provider?.getDeviceModel() ?: getDefaultDeviceModel()
            val osVersion = provider?.getOSVersion() ?: getDefaultOsVersion()
            val sdkVersion = SDKConstants.VERSION

            logger.debug("Initializing telemetry manager: device=$deviceId, model=$deviceModel, os=$osVersion")

            CppBridgeTelemetry.initialize(
                environment = environment,
                deviceId = deviceId,
                deviceModel = deviceModel,
                osVersion = osVersion,
                sdkVersion = sdkVersion,
            )

            logger.debug("Telemetry manager initialized")
        } catch (e: Exception) {
            logger.error("Failed to initialize telemetry manager: ${e.message}")
        }
    }

    /**
     * Get default device model (cross-platform fallback).
     */
    private fun getDefaultDeviceModel(): String {
        return try {
            val buildClass = Class.forName("android.os.Build")
            buildClass.getField("MODEL").get(null) as? String ?: "unknown"
        } catch (e: Exception) {
            System.getProperty("os.name") ?: "unknown"
        }
    }

    /**
     * Get default OS version (cross-platform fallback).
     */
    private fun getDefaultOsVersion(): String {
        return try {
            val versionClass = Class.forName("android.os.Build\$VERSION")
            versionClass.getField("RELEASE").get(null) as? String ?: "unknown"
        } catch (e: Exception) {
            System.getProperty("os.version") ?: "unknown"
        }
    }

    /**
     * Register the OkHttp platform HTTP transport with the C++ core.
     *
     * Installs `rac_http_transport_ops` so that every `rac_http_request_*`
     * call routes through Kotlin's [OkHttpHttpTransport]. Gives Android / JVM
     * consumers the system trust store + NetworkSecurityConfig + proxy +
     * HTTP/2 for free.
     *
     * Routed through [OkHttpHttpTransport.register] (not the raw JNI thunk)
     * so the transport's registration state — and the in-flight stream
     * registry it drains on [OkHttpHttpTransport.unregister] — stays in sync
     * with what the C++ core sees.
     *
     * Guarded: skipped silently when the native library isn't loaded ([OkHttpHttpTransport.register]
     * absorbs the UnsatisfiedLinkError) so the SDK can still boot (without
     * inference) for non-networking use cases.
     */
    private fun registerOkHttpTransport() {
        if (!CppBridgeState.nativeLibraryLoaded) {
            logger.debug("Skipping OkHttp transport registration: native lib not loaded")
            return
        }
        OkHttpHttpTransport.register()
    }

    /**
     * Unregister the OkHttp platform HTTP transport and cancel any in-flight
     * streams. Best-effort — [OkHttpHttpTransport.unregister] logs failures
     * without blocking shutdown.
     */
    private fun unregisterOkHttpTransport() {
        if (!CppBridgeState.nativeLibraryLoaded) return
        OkHttpHttpTransport.unregister()
    }

    /** Shut down native core state while the platform adapter is still valid. */
    private fun shutdownNativeCore(failures: ShutdownFailureCollector): Boolean {
        if (!CppBridgeState.nativeLibraryLoaded) return true
        val succeeded = failures.captureNativeShutdown(RunAnywhereBridge::racShutdown)
        if (!succeeded) {
            logger.warn("Native SDK shutdown failed")
        }
        return succeeded
    }

    /**
     * Try to load the native commons library.
     * This is optional - the SDK works without it for non-inference features.
     *
     * NOTE: Backend registration (LlamaCPP, ONNX) is NOT done here.
     * Apps must call LlamaCPP.register() and ONNX.register() from the
     * respective backend modules to enable AI inference.
     */
    private fun tryLoadNativeLibrary() {
        logger.debug("Starting native library loading sequence")

        CppBridgeState.nativeLibraryLoaded = RunAnywhereBridge.ensureNativeLibraryLoaded()

        if (CppBridgeState.nativeLibraryLoaded) {
            logger.info("Native commons library loaded; AI inference features available")
        } else {
            logger.warn(
                "Native commons library not available; AI inference features disabled. " + "Ensure librunanywhere_jni.so is in your APK's lib/ folder.",
            )
        }
    }

    /**
     * Phase 2: Services Initialization (Asynchronous)
     *
     * Initializes platform service adapters needed by commons Phase 2.
     *
     * Must be called after [initialize] completes.
     * Must be called from a background thread (e.g., Dispatchers.IO).
     * Mirrors Swift SDK's completeServicesInitialization()
     */
    suspend fun initializeServices() {
        // Guard: check and set initializing flag under lock, then release lock for I/O
        synchronized(lock) {
            check(!shutdownInProgress) { "CppBridge shutdown is in progress" }
            check(!shutdownRetryRequired) { "CppBridge reset is required before services initialization" }
            if (!CppBridgeState.isInitialized) {
                throw IllegalStateException("CppBridge.initialize() must be called before initializeServices()")
            }
            if (CppBridgeState.servicesInitialized || CppBridgeState.servicesInitializing) {
                return
            }
            CppBridgeState.servicesInitializing = true
        }

        try {
            // Configure the Kotlin HTTP adapter used by callback-based
            // platform services. Auth and control-plane orchestration are
            // driven by rac_sdk_init_phase2_proto through the registered
            // OkHttp transport.
            if (!HTTPClientAdapter.isConfigured) {
                val configured =
                    if (_environment == SDKEnvironment.SDK_ENVIRONMENT_DEVELOPMENT) {
                        CppBridgeDevConfig.configureHTTP()
                    } else {
                        val baseUrl = CppBridgeTelemetry.getBaseUrl()
                        val apiKey = CppBridgeTelemetry.getApiKey()
                        if (!baseUrl.isNullOrEmpty() && !apiKey.isNullOrEmpty()) {
                            HTTPClientAdapter.configure(baseUrl, apiKey)
                            true
                        } else {
                            logger.warn(
                                "HTTP adapter NOT configured: baseUrl present=${!baseUrl.isNullOrEmpty()}, " +
                                    "apiKey present=${!apiKey.isNullOrEmpty()}",
                            )
                            false
                        }
                    }
                logger.info(
                    "Phase 2 HTTP adapter configuration: configured=$configured " +
                        "(isConfigured=${HTTPClientAdapter.isConfigured})",
                )
            } else {
                logger.info("Phase 2 HTTP adapter already configured (isConfigured=true)")
            }

            SystemTTSModule.register()

            synchronized(lock) {
                CppBridgeState.servicesInitialized = true
                CppBridgeState.servicesInitializing = false
            }
            logger.debug("Phase 2 services initialization complete")
        } catch (e: Exception) {
            synchronized(lock) {
                CppBridgeState.servicesInitializing = false
            }
            throw e
        }
    }

    /**
     * Shutdown the SDK and release all resources.
     *
     * Mirrors Swift `CppBridge.shutdown()` which is async because AI component
     * destroy() methods are actor-isolated. The compatibility entry point is
     * non-suspending and awaits the same teardown via `runBlocking`; coroutine
     * callers should prefer [shutdownSuspending].
     *
     * Order (matching Swift CppBridge.shutdown() exactly):
     * 1. Destroy stateful AI component bridges (LLM → STT → TTS → VAD → VoiceAgent)
     * 2. Unregister Phase 2 services, Phase 1 core extensions in reverse order.
     *
     * Each component destroy is wrapped in try/catch so a failure in one
     * component does not abort the rest of the shutdown sequence.
     */
    fun shutdown() {
        runBlocking { shutdownSuspending() }
    }

    /** Tear down Phase 1 state after an initialization failure. */
    internal fun rollbackInitialization(afterNativeShutdown: () -> Unit = {}) {
        val failures = ShutdownFailureCollector()
        synchronized(lock) {
            if (shutdownInProgress) return
            shutdownInProgress = true
            try {
                CppBridgeDevice.unregister()
                shutdownNativeCore(failures)
                if (!runPostNativeShutdownHook(afterNativeShutdown)) {
                    failures.record("Post-native shutdown hook failed")
                }
                failures.capture("SDK event teardown failed") { CppBridgeSDKEvents.unregister() }
                failures.capture("Telemetry teardown failed") { CppBridgeTelemetry.unregister() }
                failures.capture("System TTS teardown failed") { SystemTTSModule.unregister() }
                failures.capture("HTTP transport teardown failed") { unregisterOkHttpTransport() }
                failures.capture("HTTP configuration teardown failed") { HTTPClientAdapter.reset() }
                failures.capture("Platform adapter teardown failed") { CppBridgePlatformAdapter.unregister() }

                // A failed partial rollback remains retryable through the
                // public reset path. Do not clear the bridge/native-loaded
                // gates until every cleanup step succeeds.
                if (!failures.hasFailure) {
                    failures.capture("Bridge state teardown failed") { CppBridgeState.shutdown() }
                }
                shutdownRetryRequired = failures.hasFailure
            } finally {
                shutdownInProgress = false
            }
        }
        failures.throwIfAny()
    }

    /**
     * Suspending shutdown that matches Swift `CppBridge.shutdown() async`.
     *
     * Awaits AI component actor destruction sequentially before tearing down
     * telemetry/events/platform adapter. Prefer this from already-suspending
     * callers to avoid the `runBlocking` bridge in [shutdown].
     */
    suspend fun shutdownSuspending() {
        shutdownSuspending(afterNativeShutdown = {})
    }

    /**
     * Platform-facade shutdown hook placed after the terminal native event and
     * before event and telemetry dependencies are detached.
     */
    internal suspend fun shutdownSuspending(afterNativeShutdown: () -> Unit) {
        // Claim teardown under the synchronous lifecycle lock, then release it
        // before awaiting component and telemetry work.
        synchronized(lock) {
            if ((!CppBridgeState.isInitialized && !shutdownRetryRequired) || shutdownInProgress) return
            shutdownInProgress = true
        }

        val failures = ShutdownFailureCollector()

        // Destroy AI components sequentially before tearing down Telemetry/Events.
        // Each call is best-effort: a failure in one bridge must not block the rest.
        // Matches Swift CppBridge.shutdown() ordering exactly:
        //   LLM → STT → TTS → VAD → VoiceAgent
        try {
            CppBridgeLLM.destroy()
        } catch (_: Throwable) {
            logger.warn("LLM bridge teardown failed")
            failures.record("LLM bridge teardown failed")
        }
        try {
            CppBridgeSTT.destroy()
        } catch (_: Throwable) {
            logger.warn("STT bridge teardown failed")
            failures.record("STT bridge teardown failed")
        }
        try {
            CppBridgeTTS.destroy()
        } catch (_: Throwable) {
            logger.warn("TTS bridge teardown failed")
            failures.record("TTS bridge teardown failed")
        }
        try {
            CppBridgeVAD.destroy()
        } catch (_: Throwable) {
            logger.warn("VAD bridge teardown failed")
            failures.record("VAD bridge teardown failed")
        }
        try {
            CppBridgeVoiceAgent.destroy()
        } catch (_: Throwable) {
            logger.warn("Voice-agent bridge teardown failed")
            failures.record("Voice-agent bridge teardown failed")
        }

        try {
            synchronized(lock) {
                // Stop native device callbacks before canonical shutdown begins.
                CppBridgeDevice.unregister()

                // Canonical native shutdown publishes its terminal lifecycle event
                // while EventBus, telemetry, HTTP, logging, and platform services
                // are still available.
                shutdownNativeCore(failures)
                if (!runPostNativeShutdownHook(afterNativeShutdown)) {
                    failures.record("Post-native shutdown hook failed")
                }

                if (!failures.capture("SDK event teardown failed") { CppBridgeSDKEvents.unregister() }) {
                    logger.warn("SDK event teardown failed")
                }
            }

            // Native flush callbacks run on this lifetime's replaceable scope.
            // Drain it before HTTP credentials or transport are released.
            try {
                CppBridgeTelemetry.unregisterSuspending()
            } catch (_: Throwable) {
                logger.warn("Telemetry callback drain failed during shutdown")
                failures.record("Telemetry callback drain failed during shutdown")
            }

            synchronized(lock) {
                if (!failures.capture("System TTS teardown failed") { SystemTTSModule.unregister() }) {
                    logger.warn("System TTS teardown failed")
                }

                // Telemetry flush/destroy above may use HTTP, so release the
                // transport and copied credentials only after telemetry is gone.
                if (!failures.capture("HTTP transport teardown failed") { unregisterOkHttpTransport() }) {
                    logger.warn("HTTP transport teardown failed")
                }
                if (!failures.capture("HTTP configuration teardown failed") { HTTPClientAdapter.reset() }) {
                    logger.warn("HTTP configuration teardown failed")
                }

                if (!failures.capture("Platform adapter teardown failed") { CppBridgePlatformAdapter.unregister() }) {
                    logger.warn("Platform adapter teardown failed")
                }

                // Keep the Kotlin/native-loaded gates intact after any failure
                // so a fail-closed RunAnywhere.reset() retry re-enters this
                // canonical teardown. Clear them only after every step succeeds.
                if (!failures.hasFailure &&
                    !failures.capture("Bridge state teardown failed") { CppBridgeState.shutdown() }
                ) {
                    logger.warn("Bridge state teardown failed")
                }
                shutdownRetryRequired = failures.hasFailure
            }
        } finally {
            synchronized(lock) {
                shutdownInProgress = false
            }
        }

        failures.throwIfAny()
    }

    private fun runPostNativeShutdownHook(hook: () -> Unit): Boolean =
        try {
            hook()
            true
        } catch (_: Throwable) {
            logger.warn("Post-native shutdown hook failed")
            false
        }

    /**
     * Check if the C++ core is initialized.
     *
     * @return true if rac_is_initialized() returns true
     */
    fun isNativeInitialized(): Boolean {
        if (!CppBridgeState.isInitialized || !isNativeLibraryLoaded) return false
        return RunAnywhereBridge.racIsInitialized()
    }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Foundation/Bridge/CppBridge.swift
/**
 * CppBridge.swift
 *
 * Unified bridge architecture for C++ ↔ Swift interop.
 *
 * All C++ bridges are organized under a single namespace for:
 * - Consistent initialization/shutdown lifecycle
 * - Shared access to platform resources
 * - Clear ownership and dependency management
 *
 * ## Initialization Order
 *
 * ```swift
 * // Phase 1: Core init (sync) - must be called first
 * CppBridge.initialize(environment: .production)
 *   ├─ PlatformAdapter.register()  ← File ops, logging, keychain
 *   ├─ Events.register()           ← Analytics event callback
 *   ├─ Telemetry.initialize()      ← Telemetry HTTP callback
 *   └─ Device.register()           ← Device registration callbacks
 *
 * // Phase 2: Services init (async) - after HTTP is configured
 * await CppBridge.initializeServices()
 *   └─ Platform.register()         ← LLM/TTS service callbacks
 * ```
 *
 * ## Bridge Extensions (in Extensions/ folder)
 *
 * - CppBridge+PlatformAdapter.swift - File ops, logging, keychain, clock
 * - CppBridge+Environment.swift - Environment, DevConfig, Endpoints
 * - CppBridge+Telemetry.swift - Events, Telemetry
 * - CppBridge+Device.swift - Device registration
 * - CppBridge+State.swift - SDK state management
 * - CppBridge+HTTP.swift - HTTP transport
 * - CppBridge+Auth.swift - Authentication flow
 * - CppBridge+ModelPaths.swift - Model path utilities
 * - CppBridge+ModelRegistry.swift - Model registry
 * - CppBridge+Download.swift - Download manager
 * - CppBridge+Platform.swift - Platform services (Foundation Models, System TTS)
 * - CppBridge+LLM/STT/TTS/VAD.swift - AI component bridges
 * - CppBridge+VoiceAgent.swift - Voice agent bridge
 * - CppBridge+Storage/Strategy.swift - Storage utilities
 */

import CRACommons
import Foundation
import os

// MARK: - Main Bridge Coordinator

struct CppBridgeLifetimeGate: Sendable {
    enum Phase: Sendable, Equatable {
        case idle
        case initializing
        case initialized
        case shuttingDown
    }

    enum ShutdownAdmission: Sendable, Equatable {
        case notNeeded
        case waitForInitialization
        case start
        case join
    }

    private(set) var phase: Phase = .idle

    var isInitialized: Bool { phase == .initialized }

    mutating func beginInitialization() throws -> Bool {
        switch phase {
        case .idle:
            phase = .initializing
            return true
        case .initialized:
            return false
        case .initializing, .shuttingDown:
            throw SDKException(
                code: .invalidState,
                message: "C++ bridge lifecycle transition is in progress",
                category: .internal
            )
        }
    }

    @discardableResult
    mutating func completeInitialization() -> Bool {
        guard phase == .initializing else { return false }
        phase = .initialized
        return true
    }

    mutating func beginRollback() -> Bool {
        guard phase == .initializing || phase == .initialized else { return false }
        phase = .shuttingDown
        return true
    }

    mutating func beginShutdown() -> ShutdownAdmission {
        switch phase {
        case .idle:
            return .notNeeded
        case .initializing:
            return .waitForInitialization
        case .initialized:
            phase = .shuttingDown
            return .start
        case .shuttingDown:
            return .join
        }
    }

    @discardableResult
    mutating func completeShutdown() -> Bool {
        guard phase == .shuttingDown else { return false }
        phase = .idle
        return true
    }
}

/// Central coordinator for all C++ bridges
/// Manages lifecycle and shared resources
public enum CppBridge {

    // MARK: - Shared State

    /// Combined synchronously-readable bridge state, guarded by a single
    /// `OSAllocatedUnfairLock`. Replaces the prior NSLock + 3 vars layout
    /// per AGENTS.md "Do not use NSLock as it is outdated."
    private struct CppBridgeSharedState {
        var environment: SDKEnvironment = .development
        var servicesInitialized: Bool = false
        var lifecycle = CppBridgeLifetimeGate()
        var shutdownTask: Task<Void, Never>?
    }

    private static let state =
        OSAllocatedUnfairLock<CppBridgeSharedState>(initialState: CppBridgeSharedState())

    /// Whether core bridges are initialized (Phase 1)
    public static var isInitialized: Bool {
        state.withLock { $0.lifecycle.isInitialized }
    }

    /// Whether service bridges are initialized (Phase 2)
    public static var servicesInitialized: Bool {
        state.withLock { $0.lifecycle.isInitialized && $0.servicesInitialized }
    }

    // MARK: - Phase 1: Core Initialization (Synchronous)

    /// Initialize all core C++ bridges
    ///
    /// This must be called FIRST during SDK initialization, before any C++ operations.
    /// It registers fundamental platform callbacks that C++ needs.
    ///
    /// - Parameter environment: SDK environment
    public static func initialize(environment: SDKEnvironment) throws {
        let shouldInitialize = try state.withLock { current -> Bool in
            let admitted = try current.lifecycle.beginInitialization()
            if admitted {
                current.environment = environment
            }
            return admitted
        }
        guard shouldInitialize else { return }

        do {
            // Step 1: Platform adapter FIRST (logging, file ops, keychain)
            // This must be registered before any other C++ calls
            PlatformAdapter.register()
            let deviceId = try Device.persistentId

            // Step 1.1: Register the Swift URLSession HTTP transport so
            // every subsequent `rac_http_request_*` call flows through
            // Apple's stack (trust store, ATS, proxies, HTTP/2). Must
            // happen before any other bridge that might trigger HTTP —
            // e.g. Telemetry initialization below.
            URLSessionHttpTransport.register()

            // Step 1.5: Configure C++ logging based on environment
            // In production: disables C++ stderr, logs only go through Swift bridge
            // In development: C++ stderr ON for debugging
            rac_configure_logging(environment.cEnvironment)

            // Step 1.6: Host app/client metadata. `platform` remains the OS
            // family; this separate structured payload carries the SDK binding and
            // Bundle facts for device registration.
            ClientInfo.register()

            // Step 2: Telemetry manager (builds JSON, calls HTTP callback).
            // Must come before Events.register(): the events bridge attaches this
            // manager to the C++ router as the telemetry sink, so the manager has
            // to exist first.
            Telemetry.initialize(environment: environment, deviceId: deviceId)

            // Step 3: Attach the telemetry manager as the C++ router's telemetry sink.
            Events.register()

            // Step 3.5: Start the EventBus native subscription so lifecycle/model/
            // error events flow into `EventBus.shared.events` (see EventBus.start()).
            EventBus.shared.start()

            // Step 4: Device registration callbacks
            try Device.register()

            let committed = state.withLock { $0.lifecycle.completeInitialization() }
            guard committed else {
                throw SDKException(
                    code: .invalidState,
                    message: "C++ bridge initialization lost lifecycle ownership",
                    category: .internal
                )
            }

            SDKLogger(category: "CppBridge").debug("Core bridges initialized for \(environment)")
        } catch {
            rollbackInitialization()
            throw error
        }
    }

    /// Tear down synchronous Phase-1 bridge state after a failed initialization.
    /// Safe before `isInitialized` is committed and safe to call repeatedly.
    static func rollbackInitialization() {
        let ownsRollback = state.withLock { current -> Bool in
            guard current.shutdownTask == nil else { return false }
            return current.lifecycle.beginRollback()
        }
        guard ownsRollback else { return }

        Device.unregister()
        // Canonical native shutdown publishes its lifecycle event and clears
        // Commons state while logging/event/platform dependencies still exist.
        State.shutdown()
        EventBus.shared.stop()
        Events.unregister()
        Telemetry.rollbackInitialization()
        URLSessionHttpTransport.unregister()
        PlatformAdapter.unregister()
        state.withLock {
            $0.environment = .development
            $0.servicesInitialized = false
            _ = $0.lifecycle.completeShutdown()
        }
    }

    // MARK: - Phase 2: Services Initialization (Async)

    /// Initialize service bridges that require HTTP
    ///
    /// Called after HTTP transport is configured. These bridges need
    /// network access to function.
    @MainActor
    public static func initializeServices() {
        let snapshot = state.withLock { current -> (active: Bool, alreadyDone: Bool, env: SDKEnvironment) in
            (current.lifecycle.isInitialized, current.servicesInitialized, current.environment)
        }
        guard snapshot.active, !snapshot.alreadyDone else { return }
        let currentEnv = snapshot.env

        // Model assignment fetch needs no Swift callbacks: commons routes it
        // through the registered URLSession HTTP transport, and the fetch
        // itself is owned by rac_sdk_init_phase2_proto.

        // Platform services (Foundation Models, System TTS)
        Platform.register()

        state.withLock { $0.servicesInitialized = true }

        SDKLogger(category: "CppBridge").debug("Service bridges initialized (env: \(currentEnv))")
    }

    // MARK: - Shutdown

    /// Shutdown all C++ bridges
    ///
    /// Async because AI component destroy() methods are actor-isolated.
    /// Awaiting them sequentially (instead of wrapping in `Task { ... }`)
    /// ensures Telemetry/Events teardown does not race destroy completion.
    public static func shutdown() async {
        while true {
            let decision = state.withLock { current -> ShutdownDecision in
                switch current.lifecycle.beginShutdown() {
                case .notNeeded:
                    return .complete
                case .waitForInitialization:
                    return .wait
                case .start:
                    let task = Task.detached(priority: .userInitiated) {
                        await performShutdown()
                    }
                    current.shutdownTask = task
                    return .join(task)
                case .join:
                    if let task = current.shutdownTask {
                        return .join(task)
                    }
                    return .wait
                }
            }

            switch decision {
            case .complete:
                return
            case .wait:
                await Task.yield()
            case let .join(task):
                await task.value
                return
            }
        }
    }

    private enum ShutdownDecision {
        case complete
        case wait
        case join(Task<Void, Never>)
    }

    private static func performShutdown() async {

        // Destroy AI components sequentially before tearing down Telemetry/Events
        await LLM.shared.destroy()
        await STT.shared.destroy()
        await TTS.shared.destroy()
        await VAD.shared.destroy()
        await VoiceAgent.shared.destroy()

        await MainActor.run {
            Platform.unregister()
        }

        Device.unregister()

        // Canonical native shutdown must run exactly once before its borrowed
        // platform adapter and event/log dependencies are detached.
        State.shutdown()

        // Stop the EventBus native subscription while the native ABI surface
        // is still alive (see EventBus.stop()).
        EventBus.shared.stop()

        // Detach the router's telemetry sink BEFORE destroying the manager so
        // the C++ router never holds a dangling manager pointer.
        Events.unregister()
        await Telemetry.shutdown()
        await HTTPClientAdapter.shared.shutdown()
        URLSessionHttpTransport.unregister()
        PlatformAdapter.unregister()

        state.withLock {
            $0.environment = .development
            $0.servicesInitialized = false
            $0.shutdownTask = nil
            _ = $0.lifecycle.completeShutdown()
        }

        SDKLogger(category: "CppBridge").debug("All bridges shutdown")
    }
}
